step/step.go

```diff
@@ -1,42 +1,80 @@
 package step
 
 import (
+	"encoding/json"
 	"fmt"
 	"os"
+
+	"github.com/stackpulse/steps-sdk-go/env"
 )
 
 const (
-	ExitCodeOK      = 0
-	ExitCodeFailure = 1
+	stepErrorKey = "step_error"
 )
 
 type Outputs struct {
 	Object interface{} `json:"api_object,omitempty"`
 }
 
+type Output map[string]interface{}
+
 // An interface describing a general purpose stackpulse step
 type Step interface {
-	Init() error
-	Run() (exitCode int, output []byte, err error)
+	Init() *Error
+	Run() (Output, *Error)
 }
 
-// A helper method that execute a given step according to step lifecycle
-func Run(step Step) {
-	err := step.Init()
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "failed init step arguments, %v\n", err)
-		os.Exit(ExitCodeFailure)
+var ErrSuccess = &Error{
+	Code:    OK,
+	Message: "success",
+}
+
+func printOutput(stepOutput map[string]interface{}, stepError *Error) error {
+	if stepError == nil {
+		stepError = ErrSuccess
 	}
 
-	exitCode, output, err := step.Run()
+	if stepOutput == nil {
+		stepOutput = Output{}
+	}
+
+	stepOutput[stepErrorKey] = *stepError
+
+	jsonStepOutput, err := json.Marshal(stepOutput)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Step error: %s\n", err.Error())
-		if output != nil {
-			os.Stderr.Write([]byte("\n"))
-			os.Stderr.Write(output)
+		stepError = &Error{
+			Code:    InvalidArgument,
+			Message: "something went wrong. can't parse step output",
+			Verbose: fmt.Errorf("error: %w", err).Error(),
 		}
-		os.Exit(exitCode)
+
+		stepOutput[stepErrorKey] = *stepError
+
+		return err
+	}
+
+	os.Stdout.Write([]byte(env.EndMarkerDefault + string(jsonStepOutput)))
+	return nil
+}
+
+// A helper method that execute a given step according to step lifecycle
+func Run(step Step) error {
+	if stepError := step.Init(); stepError != nil {
+		if err := printOutput(map[string]interface{}{}, stepError); err != nil {
```

Comment:
```suggestion
		if err := printOutput(nil, stepError); err != nil {
```
your function handles nil properly